alerts/plugins/ip_source_enrichment.py

```diff
@@ -0,0 +1,115 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+# Copyright (c) 2014 Mozilla Corporation
+
+from operator import add
+import os
+import re
+
+import netaddr
+
+
+CONFIG_FILE = os.path.join(
+    os.path.dirname(__file__),
+    'ip_source_enrichment.json')
+
+
+def _find_ip_addresses(string):
+    '''List all of the IPv4 and IPv6 addresses found in a string.'''
+
+    ipv4_rx = '(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)'
+    ipv6_rx = '(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9]))'
+
+    ipv4 = re.findall(ipv4_rx, string)
+    ipv6 = map(
+        lambda match: match[0] if isinstance(match, tuple) else match,
+        re.findall(ipv6_rx, string))
+
+    return ipv4 + ipv6
+
+
+def enrich(alert, known_ips):
+    '''Add information to alerts containing IP addresses that describes
+    the source location of the IP address if it can be determined based
+    on a configured mapping.
+    '''
+
+    def find_ips(value):
+        if isinstance(value, str):
+            return _find_ip_addresses(value)
+
+        if isinstance(value, list) or isinstance(value, tuple):
+            found = [find_ips(item) for item in value]
+            return reduce(add, found, [])
+
+        if isinstance(value, dict):
+            found = [find_ips(item) for item in value.values()]
+            return reduce(add, found, [])
+
+        return []
+
+    ips = find_ips(alert)
+
+    alert = alert.copy()
+
+    for ip in set(ips):
+        ip_address = netaddr.IPAddress(ip)
+
+        matching_descriptions = filter(
+            lambda known: ip_address in netaddr.IPSet([known['range']]),
+            known_ips)
+
+        for desc in matching_descriptions:
```

Comment:
What happens when both the sourceipaddress and destinationipaddress are in the known ranges? I think we would overwrite details.site with whatever came last and append both to the summary. Is this what 'should' happen?

Reply:
This is a good question and part of why I didn't add anything like `details.site` the first time around. I defer to @Phrozyn or @pwnbus or @mpurzynski regarding how we should handle the case where multiple IPs are recognized.

We could do something like:

```json
{
  "details": {
    "sites": [
      { "ip": "a.b.c.d", "site": "office1" },
      { "ip": "x.y.z.n", "site": "office2" }
    ]
  }
}
```

Reply:
This approach works for me

Reply:
I'd agree, having details.sites be an array containing all of the different ips -> office mappings.